new openshift-compatibility-gen command

sanchezl opened the pull request:
Introduce the `openshift-compatibility-gen` command do generate api compatibility level meta-data for OpenShift API types.

The methods generated provide programmatic access to OpenShift API compatibility levels (for generating reports, docs, etc.).

The generator picks up a generator tag from the type declaration.

### Type Level Tags

```go
// +openshift:compatibility-gen:level=l
// +openshift:compatibility-gen:internal
```

### Generated Comments

Human readable compatibility comments are added to types that specify the `+openshift:compatibility-gen:level=l` tag:

```go
// Compatibility Level 1: Stable within a major release for a minimum of 12 months or 3 minor releases (whichever is longer).
// Compatibility Level 2: Stable within a major release for a minimum of 9 months or 3 minor releases (whichever is longer).
// Compatibility Level 3: Will attempt to be as compatible from version to version as possible, but version to version compatibility is not guaranteed.
// Compatibility Level 4: No compatibility is provided, the API can change at any point for any reason. These capabilities should not be used by applications needing long term support.
```

The compatibility statement added to the godoc will be picked up by the openapi schema generator and eventually be displayed in the OpenShift documentation's API reference.

### Internal Types

Internal types, specifying the `+openshift:compatibility-gen:internal` tag, do no require a `+openshift:compatibility-gen:level=l` tag unless they are also exposed (i.e. have a corresponding CRD installed), in which case `+openshift:compatibility-gen:level=4` should be specified. Only `level=4` is valid for internal types.

deads2k commented:
don't duplicate these:

```
// +openshift:compatibility-gen:introduced=m.n
// +openshift:compatibility-gen:deprecated=m.n
// +openshift:compatibility-gen:removed=m.n
// +openshift:compatibility-gen:replacement=g,v,k
```

If we need them, we'll use the kube ones.

deads2k commented:
having `openshift:compatibility-gen:level=1` at the package level seems a little odd.  I'd rather have people explicitly indicate on every API they create.

deads2k commented on cmd/openshift-compatibility-gen/generators/gen_type.go:
```diff
@@ -0,0 +1,226 @@
+		return err
+	}
+
+	if versionedMethodOrDie("OpenShiftAPICompatibilityLevel", t) == nil {
```
I think you have a package name in here for this type

1. If you cannot parse out a `vN{,betaM,alphaM}`, then produce non-compiling code indicating that
2. if alpha and compatibility level not equal four, then produce non-compiling code explaining why and linking doc
3. if beta compatibility level is not two, this is not allowed.  If they choose one, they should be GA. If they choose less, they should be alpha,
4. if GA compatibility level is not one, this is not allowed.  GA APIs are supported for 12 months

sanchezl replied:
Done.

deads2k replied:
> Done.

I bet that you get more bugs about a failed generator than you do about non-compiling code (which they will check into their branhc), but I'm willing to try.

deads2k commented:
Do you have access to the full comment block?  Having a specific string in the godoc could be helpful. Something that matches each compatibility level like.

Compatibility level 1 - stable within a major release for a minimum of 12 months or 3 minor releases (whichever is longer)
Compatibility level 2 - stable within a major release for a minimum of 9 months or 3 minor releases (whichever is longer)
Compatibility level 3 - not on by default.  Will attempt to be as compatible from version to version as possible, but version to version compatibility is not guaranteed
Compatibility level 4 - not on by default.  No compatibility is provided, the API can change at any point for any reason.  These capabilities should not be used by applications needing long term support.


I'll take it as a step two if required.

sanchezl commented:
> don't duplicate these:
> 
> ```
> // +openshift:compatibility-gen:introduced=m.n
> // +openshift:compatibility-gen:deprecated=m.n
> // +openshift:compatibility-gen:removed=m.n
> // +openshift:compatibility-gen:replacement=g,v,k
> ```
> 
> If we need them, we'll use the kube ones.

Just confirming (I think you already did in another conversation) since, if we use the kube ones, they would be referring to the underlying kube versions, not openshift version explicitly.

deads2k commented on authorization/v1/types.go:
```diff
@@ -60,6 +61,7 @@ type IsPersonalSubjectAccessReview struct {
 // +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
 
 // Role is a logical grouping of PolicyRules that can be referenced as a unit by RoleBindings.
+// +openshift:compatibility-gen:level=1
```
ugh, how did we never deprecate these :(

deads2k commented on cmd/openshift-compatibility-gen/generators/gen_type.go:
```diff
@@ -0,0 +1,197 @@
+	a = a.With("level", l)
+	switch l {
+	case 1:
+		a = a.With("comment", level1Comment)
```
what does thsi do?

deads2k replied:
> what does thsi do?

Oh, I see.  That's fine, but in a followup we want to force it onto the doc for the type itself so it shows up in openapi.

deads2k commented on legacyconfig/v1/types.go:
```diff
@@ -14,6 +14,7 @@ type ExtendedArguments map[string][]string
 // +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
 
 // NodeConfig is the fully specified config starting an OpenShift node
+// +openshift:compatibility-gen:level=1
```
well, this is going to be fun.

deads2k replied:
Let's get kube deprecated values here.  We can actually mark deprecation as 4.1

deads2k commented on openshiftcontrolplane/v1/types.go:
```diff
@@ -10,6 +10,7 @@ import (
 
 // +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
 
+// +openshift:compatibility-gen:level=1
```
on-disk configuration is getting weird.  

I think we need an option for `openshift:compatibility-gen:exposed=false`, which can be paired with `openshift:compatibility-gen:level=1` for our on-disk configuration types that we don't allow customers to set.

deads2k commented on operatorcontrolplane/v1alpha1/doc.go:
```diff
@@ -4,4 +4,7 @@
 // +kubebuilder:validation:Optional
 // +groupName=controlplane.operator.openshift.io
+
+// +openshift:compatibility-gen=true
```
this is the only instance.  that's weird.

deads2k commented on cmd/openshift-compatibility-gen/generators/gen_type.go:
```diff
@@ -0,0 +1,207 @@
+	level4Comment = "No compatibility is provided, the API can change at any point for any reason. These capabilities should not be used by applications needing long term support."
+)
+
+func (g *genCompatibility) argsFromType(c *generator.Context, t *types.Type) (generator.Args, error) {
```
any chance of a unit test?

deads2k commented on kubecontrolplane/v1/types.go:
```diff
@@ -11,6 +11,8 @@ import (
 
 // +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
 
+// +openshift:compatibility-gen:level=1
```
level 4

deads2k commented on imageregistry/v1/types.go:
```diff
@@ -10,6 +10,7 @@ import (
 // +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
 
 // ConfigList is a slice of Config objects.
+// +openshift:compatibility-gen:level=1
```
level 4, exposed=false and make a note to revisit with the image team as a followup to figure  out if users touch this.

deads2k commented on imageregistry/v1/types.go:
```diff
@@ -30,6 +31,7 @@ const (
 
 // Config is the configuration object for a registry instance managed by
 // the registry operator
+// +openshift:compatibility-gen:level=1
```
level 4, exposed=false and make a note to revisit with the image team as a followup to figure  out if users touch this.

sanchezl replied:
@deads2k Looks like `Config` and `ImagePruner` (`imageregistry.operator.openshift.io` group) are exposed:

```sh-session
$ oc api-resources  --api-group=imageregistry.operator.openshift.io
NAME           SHORTNAMES   APIGROUP                              NAMESPACED   KIND
configs                     imageregistry.operator.openshift.io   false        Config
imagepruners                imageregistry.operator.openshift.io   false        ImagePruner

$ oc get configs.imageregistry.operator.openshift.io
NAME      AGE
cluster   46h

$ oc get imagepruners.imageregistry.operator.openshift.io 
NAME      AGE
cluster   46h

$ oc version
Client Version: 4.7.0-0.nightly-2020-11-30-172451
Server Version: 4.7.0-0.nightly-2020-12-04-013308
Kubernetes Version: v1.19.2+ad738ba
```

deads2k replied:
that seems... odd.  ok, thanks.

deads2k commented on imageregistry/v1/types_imagepruner.go:
```diff
@@ -12,6 +12,7 @@ import (
 // +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
 
 // ImagePrunerList is a slice of ImagePruner objects.
+// +openshift:compatibility-gen:level=1
```
level 4, exposed=false and make a note to revisit with the image team as a followup to figure  out if users touch this.

deads2k commented on imageregistry/v1/types_imagepruner.go:
```diff
@@ -24,6 +25,7 @@ type ImagePrunerList struct {
 
 // ImagePruner is the configuration object for an image registry pruner
 // managed by the registry operator.
+// +openshift:compatibility-gen:level=1
```
level 4, exposed=false and make a note to revisit with the image team as a followup to figure  out if users touch this.

deads2k commented on kubecontrolplane/v1/types.go:
```diff
@@ -187,6 +189,8 @@ type AggregatorConfig struct {
 
 // +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
 
+// +openshift:compatibility-gen:level=1
```
4

deads2k commented on legacyconfig/v1/types.go:
```diff
@@ -14,6 +14,8 @@ type ExtendedArguments map[string][]string
 // +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
 
 // NodeConfig is the fully specified config starting an OpenShift node
+// +openshift:compatibility-gen:level=1
```
4

deads2k commented on legacyconfig/v1/types.go:
```diff
@@ -186,6 +188,8 @@ type FeatureList []string
 // +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
 
 // MasterConfig holds the necessary configuration options for the OpenShift master
+// +openshift:compatibility-gen:level=1
```
4

deads2k commented on legacyconfig/v1/types.go:
```diff
@@ -821,6 +825,8 @@ type SessionConfig struct {
 // +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
 
 // SessionSecrets list the secrets to use to sign/encrypt and authenticate/decrypt created sessions.
+// +openshift:compatibility-gen:level=1
```
4

deads2k commented on legacyconfig/v1/types.go:
```diff
@@ -855,6 +861,8 @@ type IdentityProvider struct {
 // +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
 
 // BasicAuthPasswordIdentityProvider provides identities for users authenticating using HTTP basic auth credentials
+// +openshift:compatibility-gen:level=1
```
4

deads2k commented on legacyconfig/v1/types.go:
```diff
@@ -865,20 +873,26 @@ type BasicAuthPasswordIdentityProvider struct {
 // +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
 
 // AllowAllPasswordIdentityProvider provides identities for users authenticating using non-empty passwords
+// +openshift:compatibility-gen:level=1
```
you get it.  this file is all 4

deads2k commented on openshiftcontrolplane/v1/types.go:
```diff
@@ -10,6 +10,8 @@ import (
 
 // +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
 
+// +openshift:compatibility-gen:level=1
```
this whole file is 4

deads2k commented on osin/v1/types.go:
```diff
@@ -9,6 +9,8 @@ import (
 
 // +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
 
+// +openshift:compatibility-gen:level=1
```
4

deads2k commented on osin/v1/types.go:
```diff
@@ -112,6 +114,8 @@ type IdentityProvider struct {
 // +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
 
 // BasicAuthPasswordIdentityProvider provides identities for users authenticating using HTTP basic auth credentials
+// +openshift:compatibility-gen:level=1
```
4 in the whole file.

deads2k commented on security/v1/types.go:
```diff
@@ -411,6 +416,7 @@ type ServiceAccountPodSecurityPolicyReviewStatus struct {
 // +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
 
 // RangeAllocation is used so we can easily expose a RangeAllocation typed for security group
+// +openshift:compatibility-gen:level=1
```
crap, need to deprecate this one too.

deads2k commented on cmd/openshift-compatibility-gen/generators/gen_type.go:
```diff
@@ -0,0 +1,209 @@
+	beta := versionIsPrerelease(t)
+	alpha := versionIsExperimental(t)
+
+	switch {
```
this has gotten really messy, is it trying to say

```go
switch{
case ga && level != 1: fail
case beta && level !=2: fail
case alpha && level != 4: fail
case !ga && !beta && !alpha: fail
case hidden && level != 4: fail
default:
  fine
}
```

deads2k commented on hack/update-compatibility.sh:
```diff
@@ -0,0 +1,37 @@
+#!/bin/bash
+
+apiPkgs=(
```
in a followup, find a way to write this so that every subdir is selected.

deads2k commented on image/docker10/dockertypes.go:
```diff
@@ -8,6 +8,8 @@ import (
 
 // DockerImage is the type representing a container image and its various properties when
 // retrieved from the Docker client API.
+// +openshift:compatibility-gen:level=4
```
followup:  I think this is GA at a practical level.

deads2k commented on operator/v1/types_openshiftcontrollermanager.go:
```diff
@@ -9,6 +9,7 @@ import (
 // +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
 
 // OpenShiftControllerManager provides information to configure an operator to manage openshift-controller-manager.
+// +openshift:compatibility-gen:level=1
```
I think i want to start these not exposed.  the API compatibility doc indicates

> It is common practice for Operators to use Custom Resource Definitions (CRDs)  internally to accomplish a task.  These objects are not meant for use by actors external to the operator and are intended to be “hidden”.  If any CRD is not meant for use by actors external to the operator, the operators.operatorframework.io/internal-objects annotation in the Operators ClusterServiceVersion (CSV) should be specified and that signals the corresponding resource is internal use only.

https://docs.google.com/document/d/1dkfqbGfi__Cs9UZMZrv1zGaJwXoHxO84YQkMlvrfAo4/edit

let's claim that.  This can't merge until 4.8 anyway.  I think we can do this by allowing it to be exposed=false.  Perhaps a better phrasing matches the compatibility doc. `internal=true`? That defaults better.

If you can remove this package from the initial pass I think we're close enough to tag and tidy.

deads2k commented:
let's remove labeling from the internal operator types, put that in a new PR and tag the rest of the APIs to break the back of the problem.

Also, you'll probably want to start threads with console, samples, images, and builds.  I suspect they are likely to have configuration available on their individual operator resource, so they'll end up as compatibility 1 most likely.

sttts commented on apps/v1/types.go:
```diff
@@ -24,6 +24,9 @@ import (
 // is carried out and may be changed at any time. The `latestVersion` field is updated when a new deployment
 // is triggered by any means.
+//
+// Compatibility level 1: Stable within a major release for a minimum of 12 months or 3 minor releases (whichever is longer).
```
where is this sentence from? What should it tell a user? "within a major release"

sanchezl replied:
The compatibility docs define major, minor, and z-stream (_major_._minor_._z-stream_) releases.

sttts commented on apps/v1/zz_generated.openshift_compatibility.go:
```diff
@@ -0,0 +1,77 @@
+
+// CompatibilityLevel is an autogenerated function, returning the OpenShift API compatibility level.
+// It is controlled by the "openshift:compatibility-gen:level" tag in types.go.
+func (in *DeploymentConfigRollback) CompatibilityLevel() (level int) {
```
what's the plan with these methods?

sanchezl replied:
e2e tests, possibly alerts in the future.

iancoffey commented on Makefile:
```diff
@@ -52,7 +54,12 @@ update-scripts:
 	hack/update-swagger-docs.sh
 .PHONY: update-scripts
-update: update-scripts update-codegen-crds
+update: update-scripts update-compatibiity update-codegen-crds
```
s/compatibiity/compatibility/, also like 59

sanchezl replied:
Done.

sttts commented on operator/v1/types_authentication.go:
```diff
@@ -7,6 +7,8 @@ import (
 // +genclient:nonNamespaced
 // +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+// +openshift:compatibility-gen:level=4
+// +openshift:compatibility-gen:internal
```
why?

sttts replied:
For all of them in here.

sanchezl replied:
These types are internal to the operators and can change at any moment. None of them should be exposed (I am trying to see if I can [add a test](https://github.com/openshift/origin/pull/25947) for this.)

sttts replied:
They are not internal. These must be

```
// +openshift:compatibility-gen:level=1
```

sanchezl replied:
Deferring this to a followup PR. Skipping for now. Also, see: https://github.com/openshift/api/pull/803#discussion_r553551318

sttts commented on hack/update-compatibility.sh:
```diff
@@ -0,0 +1,7 @@
+inputArg=$(printf ",%s" "${apiPkgs[@]}")
+inputArg=${inputArg:1}
+
+go run github.com/openshift/api/cmd/openshift-compatibility-gen --input-dirs "$inputArg"
```
is this always running the local copy? Would have expected `go run ./cmd/....`

sanchezl replied:
I remember researching this and as long as I don't add an `@version` it should be using the local version. Unfortunately, I can't find supporting documentation for that at the moment.

sanchezl replied:
This is working as expected and running the local copy.

sttts commented on Makefile:
```diff
@@ -48,15 +51,17 @@ verify-scripts:
 	hack/verify-crds.sh
 	bash -x hack/verify-types.sh
 	hack/verify-crds-version-upgrade.sh
+	bash -x hack/verify-compatibility.sh
```
really `-x`? Maybe remove it from verify-types as well.

sttts commented:
/lgtm
/approve

Please do the rest in a follow-up.
